flink-kubernetes-operator-autoscaler/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/config/AutoScalerOptions.java

```diff
@@ -68,15 +68,16 @@ private static ConfigOptions.OptionBuilder autoScalerConfig(String key) {
     public static final ConfigOption<Double> TARGET_UTILIZATION_BOUNDARY =
             autoScalerConfig("target.utilization.boundary")
                     .doubleType()
-                    .defaultValue(0.1)
+                    .defaultValue(0.4)
```

Comment:
This makes the upper boundary to be 1.0, will this block all upscaling?

Reply:
Not at all. The boundary is used to calculate a scale down and a scale up rate. If the processing capacity falls below the scale up rate, we will scale up to reach the target capacity. If we exceeds the scale down rate, we will scale down to the target capacity. This is a bit counter-intuitive because the upper and lower bound are actually reversed.

A `1.0` utilization for the upscale threshold will lower the scale up rate which means that we delaying upscale to utilize 100% of our processing capacity based on the calculated rates. However, we will still scale up if our processing capacity is lower than the scale up rate. The scale up rate is always computed via the target rate, but the comparison is made using the actual processing capacity.

For example: Let's say we currently have a processing capacity of 100 records/sec. The processing capacity if always estimated at 100% utilization (we call this also *true rate*). At a target rate of 50 records/second (e.g. Kafka ingestion rate), the scale up bound will be 50 rec/s. That means we will only scale up once our processing capacity falls below 50 rec/s. So we will delay scaling as much as possible. If the target rate was to increase to 110 rec/s, we would scale up because our processing capacity of 100 rec/s is now lower.

Similarly, the downscale rate will be raised (instead of lowered) when we increase the utilization boundary. That means we won't scale down as quickly. The tradeoff here is slightly higher resource usage but the scaling becomes less aggressive because we will only scale down once our processing capacity exceeds the know increased "lower bound".

To illutrate this further here some sketches:

Balanced:
```
------ upscale target rate
------ processing capacity (true rate)
------ downscale target rate
```

We will scale up:
```
------ downscale rate
------ upscale rate
------ processing capacity (true rate)
```

We will scale down:

```
------ processing capacity (true rate)
------ downscale rate
------ upscale rate
```

Reply:
I see your point. But I have to admit that it is counter-intuitive and the option description "Target vertex utilization boundary. Scaling won't be performed if utilization is within (target - boundary, target + boundary)" is rather misleading

Reply:
True, the description suggests that the scaling is simply based on CPU utilization. I can try to rephrase.

Reply:
Rephrased.

Reply:
@mxm your illustration was helpful. I'm having a hard time understanding why the autoscaler makes the decisions it makes. The logs, metrics, and definitions of terms are somewhat vague (e.g. I never knew what was meant by "true processing rate" until I saw you explain it here).

I can open a new ticket to document and perhaps work on it myself, because I think these are critical definitions 😄 

But is source backlog considered in this as well? I have a source which is lagging quite substantially, but never seems to scale out because the "true processing rate" suggests it should be keeping up, but it never does. The catch-up duration is having no effect, either.

At the root of what I am asking is: even if the "true" rate is _calculated_ to be sufficient to catch up, _but it isn't actually catching up_ (let's say because a source is doing something inefficient with its kafka poll settings)... is there any mechanism in place to detect this?

Reply:
Thanks, that makes a lot of sense! Is catchup data rate determined by literal timestamps compared against the catchup duration? eg if a record was placed in kafka 10m ago, and our expected catchup duration is 5m, then are we 5m behind, or are we still 10m behind? just trying to get a better sense of "catch up" statistics!

Perhaps our problem is that lag, for every single job tracked (operator 1.7, Flink 1.18.1, all using `KafkaSource`), is `N/A`. At least according to the exposed operator metrics themselves. If the operator can't see the lag then maybe it can't make an informed decision? I'm wondering if this is a bug on our configuration or maybe I'm just way off base. I should expect to see values for `LAG_Current`, right?

Reply:
lag time is defined as: number of outstanding records / current processing rate (so it's not based on timestamps)
I would double check the pendingRecords metrics from the Kafka source because those should be there

Reply:
Great, thank you!

Yes pendingRecords is present in all jobs and has a value (fluctuating, of course). I have never seen the operator's metric `jobVertexID_LAG_Current` be anything other than N/A, though. I'm observing it via prometheus, so maybe it's just a bug in the translation layer.

(also, if this conversation is out of scope I'd be happy to move it to somewhere that is less tangential!)

Reply:
Just checked in the operator logs as well, to potentially rule out a prom translation issue. I only see `NaN` for all three metrics in the logs.